File: src/db-migration-service/src/main/java/com/cgtfarmer/demo/accessor/LambdaParameterSecretClient.java
package com.cgtfarmer.demo.accessor;

import com.amazonaws.services.lambda.runtime.LambdaLogger;
import com.cgtfarmer.demo.dto.SecretsManagerResponse;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;

public class LambdaParameterSecretClient {

  private static String tokenHeaderKey = "X-Aws-Parameters-Secrets-Token";

  private static String secretAccessUrl = "http://localhost:2773/secretsmanager/get";

  public HttpClient httpClient;

  public ObjectMapper mapper;

  public LambdaLogger logger;

  public LambdaParameterSecretClient(LambdaLogger logger, HttpClient httpClient, ObjectMapper mapper) {
    this.logger = logger;
    this.httpClient = httpClient;
    this.mapper = mapper;
  }

  public SecretsManagerResponse getSecret(String awsSessionToken, String secretId)
      throws InterruptedException, IOException {

    String url = String.format("%s?secretId=%s", secretAccessUrl, secretId);

    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(url))
        .setHeader(tokenHeaderKey, awsSessionToken)
        .build();

    HttpResponse<String> response = this.httpClient.send(
      request,
      BodyHandlers.ofString()
    );

    String body = response.body();

    this.logger.log("Secret Response Body:");
    this.logger.log(body);

    SecretsManagerResponse mappedResponse = mapper.readValue(body, SecretsManagerResponse.class);

    return mappedResponse;
  }
}

File: src/user-service/accessor/lambda-parameter-secret-client.ts
import SecretsManagerResponse from '../dto/secrets-manager-response';

export default class LambdaParameterSecretClient {

  private static tokenHeaderKey = 'X-Aws-Parameters-Secrets-Token';

  private static secretAccessUrl = 'http://localhost:2773/secretsmanager/get';

  public readonly headers: Headers;

  constructor(awsSessionToken: string) {
    this.headers = new Headers();

    this.headers.append(LambdaParameterSecretClient.tokenHeaderKey, awsSessionToken);
  }

  public async getSecret(secretId: string) {
    const url = `${LambdaParameterSecretClient.secretAccessUrl}?secretId=${secretId}`;

    const results = await fetch(url, {
      method: 'GET',
      headers: this.headers,
    });

    const body: SecretsManagerResponse = await results.json();

    return body;
  }
}
